src/Iterators.jl

```diff
@@ -311,16 +311,13 @@ if VERSION >= v"0.4-dev"
 else
     eltype(p::Product) = tuple(map(eltype, p.xss)...)
 end
-length(p::Product) = prod(map(length, p.xss))
+length(p::Product) = isempty(p.xss) ? 1 : prod(map(length, p.xss))
```

Comment:
This can be changed into `mapreduce(length, *, 1, x)` (which is orders of magnitude faster for some reason)